### PR TITLE
Bugfix/bb 227/address cold storage status archive task issues

### PR DESCRIPTION
**Pull request by alexanderchan-scality**
addresses tickets:
* https://scality.atlassian.net/browse/BB-222
* https://scality.atlassian.net/browse/BB-224
* https://scality.atlassian.net/browse/BB-225
* https://scality.atlassian.net/browse/BB-226
* https://scality.atlassian.net/browse/BB-229

### CODE REVIEW

File: extensions/lifecycle/objectProcessor/LifecycleObjectProcessor.js [hunk 1]
```diff
@@ -185,7 +185,10 @@ class LifecycleObjectProcessor extends EventEmitter {
 
         const actionEntry = ActionQueueEntry.createFromKafkaEntry(kafkaEntry);
         if (actionEntry.error) {
-            this._log.error('malformed action entry', kafkaEntry.value);
+            this._log.error('malformed action entry', {
+                error: actionEntry.error,
+                entry: kafkaEntry.value,
+            });
```
Comment (alexanderchan-scality) [inline extensions/lifecycle/objectProcessor/LifecycleObjectProcessor.js:197]: fixed: https://github.com/scality/backbeat/pull/2254#discussion_r914160871

File: extensions/lifecycle/objectProcessor/LifecycleObjectTransitionProcessor.js [hunk 2]
```diff
@@ -99,7 +98,10 @@ class LifecycleObjectTransitionProcessor extends LifecycleObjectProcessor {
     processColdStorageStatusEntry(kafkaEntry, done) {
         const entry = ColdStorageStatusQueueEntry.createFromKafkaEntry(kafkaEntry);
         if (entry.error) {
-            this._log.error('malformed status entry', kafkaEntry.value);
+            this._log.error('malformed status entry', {
+                error: entry.error,
+                entry: kafkaEntry.value,
+            });
```
Comment (alexanderchan-scality) [inline extensions/lifecycle/objectProcessor/LifecycleObjectTransitionProcessor.js:111]: fixed: https://github.com/scality/backbeat/pull/2254#discussion_r914160871

**Review by miniscruff — APPROVED** [review]

**Review by naren-rajendran — APPROVED** [review]

**Comment by rachedbenmustapha** [comment]
@alexanderchan-scality  I ran a test, seeing BB-229 partially fixed: `archive` is populated, batch delete is issued, `location` is set to `null`. However `dataStoreName` is still `us-east-1`.

**Comment by alexanderchan-scality** [comment]
@naren-rajendran @miniscruff 
added a new commit to update `dataStoreName` which was missing: https://github.com/scality/backbeat/pull/2256/commits/1ec1b56d5aaeef9a33ddb0fd2c4e91067c36cede

File: extensions/lifecycle/objectProcessor/LifecycleObjectTransitionProcessor.js [hunk 3]
```diff
@@ -41,17 +39,30 @@ class LifecycleObjectTransitionProcessor extends LifecycleObjectProcessor {
...
     getConsumerParams() {
         const consumerParams = super.getConsumerParams();
+        const locations = require('../../../conf/locationConfig.json') || {};
 
         this._lcConfig.coldStorageTopics.forEach(topic => {
+            if (!topic.startsWith(this._lcConfig.coldStorageStatusTopicPrefix)) {
+                return;
+            }
+
+            const coldLocation =
```
Comment (naren-rajendran) [inline extensions/lifecycle/objectProcessor/LifecycleObjectTransitionProcessor.js]:
```suggestion
            const coldLocation = topic.slice(this._lcConfig.coldStorageStatusTopicPrefix.length);
```

File: extensions/lifecycle/objectProcessor/LifecycleObjectTransitionProcessor.js [hunk 4]
```diff
@@ -41,17 +39,30 @@ class LifecycleObjectTransitionProcessor extends LifecycleObjectProcessor {
...
+                topic.slice(this._lcConfig.coldStorageStatusTopicPrefix.length);
+
+            assert(locations[coldLocation],
```
Comment (naren-rajendran) [inline extensions/lifecycle/objectProcessor/LifecycleObjectTransitionProcessor.js]:
```suggestion
            assert(locations[coldLocation], `${coldLocation}: unknown location`);
```

File: extensions/lifecycle/objectProcessor/LifecycleObjectTransitionProcessor.js [hunk 5]
```diff
@@ -41,17 +39,30 @@ class LifecycleObjectTransitionProcessor extends LifecycleObjectProcessor {
...
+                `${coldLocation}: unknown location`);
+
+            assert(locations[coldLocation].isCold,
```
Comment (naren-rajendran) [inline extensions/lifecycle/objectProcessor/LifecycleObjectTransitionProcessor.js]: same here.

File: extensions/lifecycle/objectProcessor/LifecycleObjectTransitionProcessor.js (hunk 3 above)
Comment (naren-rajendran) [reply extensions/lifecycle/objectProcessor/LifecycleObjectTransitionProcessor.js]: column length is increased to 120, should be safe here.

File: extensions/lifecycle/objectProcessor/LifecycleObjectTransitionProcessor.js [hunk 6]
```diff
@@ -64,7 +75,8 @@ class LifecycleObjectTransitionProcessor extends LifecycleObjectProcessor {
                 topic,
                 groupId: this._processConfig.groupId,
                 concurrency: this._processConfig.concurrency,
-                queueProcessor: this.processColdStorageStatusEntry.bind(this),
+                queueProcessor:
+                    this.processColdStorageStatusEntry.bind(this, coldLocation),
```
Comment (naren-rajendran) [inline extensions/lifecycle/objectProcessor/LifecycleObjectTransitionProcessor.js]: same.

**Comment by nicolas2bert** [comment]
Tested the new change, and it looks good for `dataStoreName`, however, to avoid transitioning again, we check for the object's metadata storage class: https://github.com/scality/Arsenal/blob/development/8.1/lib/s3middleware/lifecycleHelpers/LifecycleUtils.ts#L245

Note: It looks like we update the metadata storage class when doing a non-cold transition: https://github.com/scality/backbeat/blob/development/8.4/extensions/lifecycle/tasks/LifecycleUpdateTransitionTask.js#L74 
```
{ "_id" : "n2b1", "value" : { "owner-display-name" : "account1", "owner-id" : "56a321265db286f6045e24cadae44a9f3874dd8deca559a16718f268031d8b18", "cache-control" : "", "content-disposition" : "", "content-encoding" : "", "expires" : "", "content-length" : 307, "content-type" : "", "content-md5" : "976834e7eb1af689962cec5b38346751", "content-language" : "", "creation-time" : "2022-07-07T13:25:45.277Z", "x-amz-version-id" : "null", "x-amz-server-version-id" : "", "x-amz-storage-class" : "STANDARD", "x-amz-server-side-encryption" : "", "x-amz-server-side-encryption-aws-kms-key-id" : "", "x-amz-server-side-encryption-customer-algorithm" : "", "x-amz-website-redirect-location" : "", "acl" : { "Canned" : "private", "FULL_CONTROL" : [ ], "WRITE_ACP" : [ ], "READ" : [ ], "READ_ACP" : [ ] }, "key" : "n2b1", "location" : null, "isDeleteMarker" : false, "versionId" : "98342799654721999999RG001  57567.1", "tags" : {  }, "replicationInfo" : { "status" : "", "backends" : [ ], "content" : [ ], "destination" : "", "storageClass" : "", "role" : "", "storageType" : "", "dataStoreVersionId" : "", "isNFS" : null }, "dataStoreName" : "location-dmf-v1", "originOp" : "s3:ObjectCreated:Put", "last-modified" : "2022-07-07T13:25:45.277Z", "md-model-version" : 5, "archive" : { "archiveInfo" : { "archiveId" : "6e93268c-8516-4018-a2c8-5d6e5ee9d1c5", "archiveVersion" : 5428658603350578 } } } }
```

**Comment by alexanderchan-scality** [comment]
> Tested the new change, and it looks good for `dataStoreName`, however, to avoid transitioning again, we check for the object's metadata storage class: https://github.com/scality/Arsenal/blob/development/8.1/lib/s3middleware/lifecycleHelpers/LifecycleUtils.ts#L245
> 
> Note: It looks like we update the metadata storage class when doing a non-cold transition: https://github.com/scality/backbeat/blob/development/8.4/extensions/lifecycle/tasks/LifecycleUpdateTransitionTask.js#L74
> 
> ```
> { "_id" : "�n2b1", "value" : { "owner-display-name" : "account1", "owner-id" : "56a321265db286f6045e24cadae44a9f3874dd8deca559a16718f268031d8b18", "cache-control" : "", "content-disposition" : "", "content-encoding" : "", "expires" : "", "content-length" : 307, "content-type" : "", "content-md5" : "976834e7eb1af689962cec5b38346751", "content-language" : "", "creation-time" : "2022-07-07T13:25:45.277Z", "x-amz-version-id" : "null", "x-amz-server-version-id" : "", "x-amz-storage-class" : "STANDARD", "x-amz-server-side-encryption" : "", "x-amz-server-side-encryption-aws-kms-key-id" : "", "x-amz-server-side-encryption-customer-algorithm" : "", "x-amz-website-redirect-location" : "", "acl" : { "Canned" : "private", "FULL_CONTROL" : [ ], "WRITE_ACP" : [ ], "READ" : [ ], "READ_ACP" : [ ] }, "key" : "n2b1", "location" : null, "isDeleteMarker" : false, "versionId" : "98342799654721999999RG001  57567.1", "tags" : {  }, "replicationInfo" : { "status" : "", "backends" : [ ], "content" : [ ], "destination" : "", "storageClass" : "", "role" : "", "storageType" : "", "dataStoreVersionId" : "", "isNFS" : null }, "dataStoreName" : "location-dmf-v1", "originOp" : "s3:ObjectCreated:Put", "last-modified" : "2022-07-07T13:25:45.277Z", "md-model-version" : 5, "archive" : { "archiveInfo" : { "archiveId" : "6e93268c-8516-4018-a2c8-5d6e5ee9d1c5", "archiveVersion" : 5428658603350578 } } } }
> ```

updated PR to also set `amzStorageClass` to cold location

**Comment by alexanderchan-scality** [comment]
ping

**Pull request merged**